### PR TITLE
Signup use case backend. 

### PR DESCRIPTION
**Pull request by stefanateljevic04**
All of the necessary files / clean architecture components of the signup use case. I'll wait for Vrinda's pull request to be approved before adding changes to main. @vrindasubhash @jackkfan0305 can you review this?

### CODE REVIEW

**Review by vrindasubhash — COMMENTED** [review]
Coding styles looks good, except for naming conventions for the use_case directory there seems to be an underscore "use_case_signup", I believe it should look like "use_case/signup", I think you can fix this by making signup a folder in the use_case folder. There does not seem to be any errors, but we can test it when we have enough code such that it compiles. Adheres to CA and SOLID. Looks good stef!

**Comment by jackkfan0305** [comment]
Adding onto what @vrindasubhash said, you should create packages under our pre-existing packages. For example, under ```interface_adapter```, you should create another package called ```signup``` and put everything related to interface_adapter and so on. This keeps everything orgainzed and prevents merging issues. Other than that, everything looks good!

File: src/interface_adapter/signup/SignupController.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+//
+// Source code recreated from a .class file by IntelliJ IDEA
+// (powered by FernFlower decompiler)
+//
```
Comment (merrickliu888) [inline src/interface_adapter/signup/SignupController.java:4]: What is this? Could we remove it?

File: src/use_case_signup/SignupInputBoundary.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+//
+// Source code recreated from a .class file by IntelliJ IDEA
+// (powered by FernFlower decompiler)
+//
+
+package use_case_signup;
+
+public interface SignupInputBoundary {
+    void execute(SignupInputData var1);
```
Comment (merrickliu888) [inline src/use_case_signup/SignupInputBoundary.java:9]: For readability purposes, could you change this to a more informative name, like `signupInputData`.

File: src/use_case_signup/SignupInputData.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// Source code recreated from a .class file by IntelliJ IDEA
+// (powered by FernFlower decompiler)
+//
+
+package use_case_signup;
+
+public class SignupInputData {
+    private final String username;
+    private final String password;
+    private final String repeatPassword;
```
Comment (merrickliu888) [inline src/use_case_signup/SignupInputData.java:11]: Keep the variable names consistent. Above you call it `password1` and `password2`. I think `repeatPassword` is more informative, so if you could change the name of the password variables to be consistent, that would make readability better.

File: src/use_case_signup/SignupInteractor.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    final UserFactory userFactory;
+
+    public SignupInteractor(SignupUserDataAccessInterface signupDataAccessInterface, SignupOutputBoundary signupOutputBoundary, UserFactory userFactory) {
+        this.userDataAccessObject = signupDataAccessInterface;
+        this.userPresenter = signupOutputBoundary;
+        this.userFactory = userFactory;
+    }
+
+    public void execute(SignupInputData signupInputData) {
+        if (this.userDataAccessObject.existsByName(signupInputData.getUsername())) {
```
Comment (merrickliu888) [inline src/use_case_signup/SignupInteractor.java:24]: Change the methods calls here to the FileDataAccessObject to be consistent with the ones in the FileDataAccessObject.

File: src/use_case_signup/SignupOutputData.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    public String getUsername() {
+        return this.username;
+    }
+
+    public String getCreationTime() {
+        return this.creationTime;
+    }
+
+    public void setCreationTime(String creationTime) {
```
Comment (merrickliu888) [inline src/use_case_signup/SignupOutputData.java:27]: We don't need anything regarding creationTime so remove it.

File: src/use_case_signup/SignupUserDataAccessInterface.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+// Source code recreated from a .class file by IntelliJ IDEA
+// (powered by FernFlower decompiler)
+//
+
+package use_case_signup;
+
+import entity.User;
+
+public interface SignupUserDataAccessInterface {
+    boolean existsByName(String var1);
```
Comment (merrickliu888) [inline src/use_case_signup/SignupUserDataAccessInterface.java:11]: Use more informative variable names.

**Review by merrickliu888 — CHANGES_REQUESTED** [review]
Overall good start. Here are some suggestions:
- Separate the view, viewmodel, state, presenter classes into another PR. 
- Remove all the comments on the top of the file mentioning the decompiler.
- Make sure your calls to the FileDataAccessObject have the right method names.
- Make the variable naming scheme more informative and consistent.
- Remove anything with saving the time the user was created. We are not using it.

File: src/use_case_signup/SignupUserDataAccessInterface.java (hunk 6 above)
Comment (merrickliu888) [inline src/use_case_signup/SignupUserDataAccessInterface.java:11]: Change this function signature to `boolean userExists(String username)`

**Review by merrickliu888 — CHANGES_REQUESTED** [review]
File DAO changes to match function signatures of current implementation.

File: src/interface_adapter/signup/SignupPresenter.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    public SignupPresenter(ViewManagerModel viewManagerModel, SignupViewModel signupViewModel, LoginViewModel loginViewModel) {
+        this.viewManagerModel = viewManagerModel;
+        this.signupViewModel = signupViewModel;
+        this.loginViewModel = loginViewModel;
+    }
+
+    public void prepareSuccessView(SignupOutputData response) {
+        LocalDateTime responseTime = LocalDateTime.parse(response.getCreationTime());
+        response.setCreationTime(responseTime.format(DateTimeFormatter.ofPattern("hh:mm:ss")));
```
Comment (samaslanowicz) [inline src/interface_adapter/signup/SignupPresenter.java:29]: I don't think we need to show creation time or response time, you could probably remove these two lines

File: src/use_case_signup/SignupInteractor.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+            this.userPresenter.prepareFailView("User already exists.");
+        } else if (!signupInputData.getPassword().equals(signupInputData.getRepeatPassword())) {
+            this.userPresenter.prepareFailView("Passwords don't match.");
+        } else {
+            LocalDateTime now = LocalDateTime.now();
```
Comment (samaslanowicz) [inline src/use_case_signup/SignupInteractor.java:29]: Here too, could probably just remove the creationTime variable from the SignUpOutputData Class

**Comment by samaslanowicz** [comment]
I also just added a few comments on the files, but seems like Merrick already covered most of it. Other than that, looks good to me! Probably ready to merge in once you make some of those small edits.

**Comment by stefanateljevic04** [comment]
@merrickliu888 @samaslanowicz I made the changes and committed, thanks for the feedback

**Review by merrickliu888 — APPROVED** [review]
Looks good.

File: src/interface_adapter/signup/SignupController.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+import use_case_signup.SignupInputData;
+
+public class SignupController {
+    final SignupInputBoundary userSignupUseCaseInteractor;
+
+    public SignupController(SignupInputBoundary userSignupUseCaseInteractor) {
+        this.userSignupUseCaseInteractor = userSignupUseCaseInteractor;
+    }
+
+    public void execute(String username, String password, String repeatPassword) {
```
Comment (samaslanowicz) [inline src/interface_adapter/signup/SignupController.java:14]: should add a javadoc for this method

File: src/use_case_signup/SignupInteractor.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    final SignupOutputBoundary userPresenter;
+    final UserFactory userFactory;
+
+    public SignupInteractor(SignupDataAccessInterface signupDataAccessInterface, SignupOutputBoundary signupOutputBoundary, UserFactory userFactory) {
+        this.userDataAccessObject = signupDataAccessInterface;
+        this.userPresenter = signupOutputBoundary;
+        this.userFactory = userFactory;
+    }
+
+    public void execute(SignupInputData signupInputData) {
```
Comment (samaslanowicz) [inline src/use_case_signup/SignupInteractor.java:18]: javadoc here too. should be added to any public methods

**Pull request merged**